File: src/hivemind/plugin/hooks/hv-pre-commit.js
// Claude Code PreToolUse hook
// Intercepts Bash tool calls containing "git commit"
// Injects spec sync reminders when .hivemind-link.json exists
//
// Input (stdin):  JSON with { tool_name, tool_input }
// Output (stdout): JSON with { status, additionalContext? }

const fs = require("fs");
const path = require("path");
const { execSync } = require("child_process");

function main() {
  // Read JSON from stdin
  let input;
  try {
    const raw = fs.readFileSync(0, "utf-8");
    input = JSON.parse(raw);
  } catch {
    // If stdin is empty or invalid JSON, approve by default
    process.stdout.write(JSON.stringify({ status: "approve" }));
    return;
  }

  const toolName = input.tool_name;
  const toolInput = input.tool_input || {};

  // 1. If not a Bash tool call, approve immediately
  if (toolName !== "Bash") {
    process.stdout.write(JSON.stringify({ status: "approve" }));
    return;
  }

  // 2. If the command doesn't contain "git commit", approve
  const command = toolInput.command || "";
  if (!command.includes("git commit")) {
    process.stdout.write(JSON.stringify({ status: "approve" }));
    return;
  }

  // 3. Check if .hivemind-link.json exists in cwd
  const linkPath = path.join(process.cwd(), ".hivemind-link.json");
  if (!fs.existsSync(linkPath)) {
    process.stdout.write(JSON.stringify({ status: "approve" }));
    return;
  }

  // 4. Read .hivemind-link.json to get the project name
  let linkData;
  try {
    linkData = JSON.parse(fs.readFileSync(linkPath, "utf-8"));
  } catch {
    // If file is malformed, approve without reminder
    process.stdout.write(JSON.stringify({ status: "approve" }));
    return;
  }

  const projectName = linkData.project || "unknown";

  // 5. Get staged files via git diff --cached --name-only
  let stagedFiles;
  try {
    const output = execSync("git diff --cached --name-only", {
      encoding: "utf-8",
      timeout: 5000,
    });
    stagedFiles = output.trim();
  } catch {
    // If git command fails, approve without reminder
    process.stdout.write(JSON.stringify({ status: "approve" }));
    return;
  }

  if (!stagedFiles) {
    process.stdout.write(JSON.stringify({ status: "approve" }));
    return;
  }

  // 6. Build reminder message
  const reminder =
    "Remember to update harness specs in projects/" +
    projectName +
    "/ if these code changes affect documented architecture or features: " +
    stagedFiles.split("\n").join(", ");

  process.stdout.write(
    JSON.stringify({ status: "approve", additionalContext: reminder })
  );
}

main();

File: src/hivemind/plugin/hooks/hv_pre_commit.py
#!/usr/bin/env python3
"""PreToolUse hook (Bash matcher).

Intercepts ``git commit`` calls. When the cwd contains a hivemind link file
(v5: ``hivemind/link.json``; legacy: ``.hivemind-link.json``), injects a
spec-sync reminder listing the staged files so the agent remembers to update
harness docs.

Input:  JSON on stdin: ``{tool_name, tool_input, cwd, ...}``
Output: JSON on stdout: ``{status: "approve", additionalContext?}``
"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path


def _approve(context: str | None = None) -> None:
    payload: dict[str, object] = {"status": "approve"}
    if context:
        payload["additionalContext"] = context
    sys.stdout.write(json.dumps(payload))


def _read_input() -> dict[str, object] | None:
    try:
        raw = sys.stdin.read()
        return json.loads(raw) if raw else None
    except (OSError, json.JSONDecodeError):
        return None


def main() -> None:
    data = _read_input()
    if data is None:
        _approve()
        return

    if data.get("tool_name") != "Bash":
        _approve()
        return

    tool_input = data.get("tool_input") or {}
    if not isinstance(tool_input, dict):
        _approve()
        return
    command = str(tool_input.get("command", ""))
    if "git commit" not in command:
        _approve()
        return

    cwd = Path(str(data.get("cwd") or Path.cwd()))
    link_file = cwd / "hivemind" / "link.json"
    if not link_file.exists():
        legacy = cwd / ".hivemind-link.json"
        if not legacy.exists():
            _approve()
            return
        link_file = legacy

    try:
        link = json.loads(link_file.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        _approve()
        return

    project = link.get("project") or "unknown"

    try:
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only"],
            cwd=str(cwd),
            capture_output=True,
            text=True,
            timeout=5,
            check=False,
        )
    except (OSError, subprocess.SubprocessError):
        _approve()
        return

    staged = result.stdout.strip()
    if not staged:
        _approve()
        return

    file_list = ", ".join(staged.splitlines())
    reminder = (
        f"Remember to update harness specs in projects/{project}/ "
        f"if these code changes affect documented architecture or features: "
        f"{file_list}"
    )
    _approve(reminder)


if __name__ == "__main__":
    main()
